### ISSUE TITLE
Rescoring of cif files generated by fpocket fails

### ISSUE THREAD

**Issue by Yorick-126** [rdk/p2rank#72]
Hi rdk!

I am having an issue with rescoring mmcif files put out by fpocket4. Rescoring here does work for pdb files generated by fpocket, For cif files, however, prank generates an empty rescoring csv. For example this is the content of the rescoring csv for protein [A0A0H2XEI7](https://alphafold.ebi.ac.uk/entry/A0A0H2XEI7), rescored on the cif file:
```
name,score,rank,old_rank,change,

```

versus the same protein but then rescored on the pdb file:
```
name,score,rank,old_rank,change,   
pocket.4,10.4213,1,4,3,+++ 
pocket.14,5.7439,2,14,12,+++++++++++++ 
pocket.3,4.7987,3,3,0, 
pocket.1,4.4184,4,1,-3,--- 
pocket.11,3.2813,5,11,6,++++++ 
pocket.7,3.1858,6,7,1,+ 
pocket.8,2.008,7,8,1,+ 
pocket.2,1.8147,8,2,-6,------ 
pocket.9,1.7731,9,9,0, 
pocket.12,1.6623,10,12,2,++ 
pocket.10,1.4871,11,10,-1,- 
pocket.5,1.3839,12,5,-7,------- 
pocket.13,1.357,13,13,0, 
pocket.6,1.0156,14,6,-8,--------- 
pocket.15,0.5874,15,15,0, 

```

I traced it down to the fpocketloader groovy: src/main/groovy/cz/siret/prank/domain/loaders/pockets/FPocketLoader.groovy.
At line 300 the function `private Structure loadPocketStructureAndDetailsFromCif` seems not be implemented because you mention at line 252 that 
` * ! for now only pdb works due to biojava cif parsing error`. So that seems to be an mmcif parsing problem on biojava's side.

Related, I read in issue [54](https://github.com/rdk/p2rank/issues/54) that you did a temporary fix for the pymol cif export. Furthermore, I read in the biojava issue (#[775](https://github.com/biojava/biojava/issues/775)) you referenced in issue 54, that the problem seems to be resolved for the pymol mmcif export on biojava's side. 

I am not sure if that would mean that biojava now correctly parses mmcif files, and if that then would work for rescoring fpocket mmcif files, but I thought you might have an idea.

So would you have an update on this part of the code? Would help me greatly!

Thanks a lot!
Yorick

**Comment by rdk** [comment]
Hi @Yorick-126 

thanks for the bug report. I'm actually working on rescoring now and plan to release a minor version with more robust loading of fpocket predictions soon, hopefully in few days. You did a great job tracking this down and reminding me what I wanted to do! Thanks!

R.

**Comment by Yorick-126** [comment]
That sounds great, looking forward to your update!

**Comment by rdk** [comment]
@Yorick-126 I have implemented support for rescoring fpocket cif predictions.

If you'd like to beta-test it you can build it from sources (develop branch) or download a build here https://drive.google.com/file/d/14_cdHsogaMck5V1sIWQXI9iSGezaQPvG/view?usp=drive_link

What kind of structures are you working with: from PDB, AlphaFold predictions or something else? I'm asking because the current rescoring model is using b-factor data from structure files as an input feature. There are many types of structures for which this is not meaningful (basically anything except x-ray structures) and it can lead to sub-optimal rescoring peformance (although in my experiance it's only slightly worse). I do plan to train new rescoring models with new major release that won't use b-factor. Let me know if you need it now.

**Comment by Yorick-126** [comment]
@rdk I did some tests with some of the cif files I have and it all seems smooth sailing! For  [A0A0H2XEI7](https://alphafold.ebi.ac.uk/entry/A0A0H2XEI7), the PDB and Cif rescoring is identical. So thanks a lot for implementing!

Concerning your second question, I mainly work with Alphafold models at the moment, sometimes PDB as well, but I would definitely be interested in rescoring models that do not use the b-factor column. I already noticed that there is an alphafold configuration file for pocket predicition. So, I tried that as well on the pocket rescoring on my pockets and that does generate different results. But do I understand correctly that this is unexpected behaviour then?

For example for [A0A0H2XEI7](https://alphafold.ebi.ac.uk/entry/A0A0H2XEI7), the following commands rescore its pockets and generate results with different scores and a slightly different pocket order:
alphafold config: `./prank rescore cif_to_cif/cif_rescore.ds -o cif_to_cif/beta_build_test/alphafold -c ./config/alphafold`
non-alphafold config: `./prank rescore cif_to_cif/cif_rescore.ds -o cif_to_cif/beta_build_test/non_alphafold`

rescoring with alphafold config results:
```
name,score,rank,old_rank,change,   
pocket.4,10.4213,1,4,3,+++ 
pocket.14,5.7439,2,14,12,+++++++++++++ 
pocket.3,4.7987,3,3,0, 
pocket.1,4.4184,4,1,-3,--- 
pocket.11,3.2813,5,11,6,++++++ 
pocket.7,3.1858,6,7,1,+ 
pocket.8,2.008,7,8,1,+ 
pocket.2,1.8147,8,2,-6,------ 
pocket.9,1.7731,9,9,0, 
pocket.12,1.6623,10,12,2,++ 
pocket.10,1.4871,11,10,-1,- 
pocket.5,1.3839,12,5,-7,------- 
pocket.13,1.357,13,13,0, 
pocket.6,1.0156,14,6,-8,--------- 
pocket.15,0.5874,15,15,0,
 ```
versus rescoring without alphafold config results:
```
name,score,rank,old_rank,change,   
pocket.4,5.0982,1,4,3,+++ 
pocket.14,2.4899,2,14,12,+++++++++++++ 
pocket.3,0.5405,3,3,0, 
pocket.7,0.2776,4,7,3,+++ 
pocket.9,0.2234,5,9,4,++++ 
pocket.8,0.2209,6,8,2,++ 
pocket.5,0.2166,7,5,-2,-- 
pocket.1,0.1851,8,1,-7,------- 
pocket.12,0.1574,9,12,3,+++ 
pocket.13,0.0902,10,13,3,+++ 
pocket.11,0.0725,11,11,0, 
pocket.15,0.0601,12,15,3,+++ 
pocket.2,0.0411,13,2,-11,------------ 
pocket.6,0.0263,14,6,-8,--------- 
pocket.10,0.0222,15,10,-5,----- 
```` 


Lastly, off-note, it seems that p2rank doesn't support java 11 anymore? I had to upgrade to java17

Again, thanks a lot!
Yorick

**Comment by rdk** [comment]
I've released the version 2.4.2 with support fo re-scoring fpocket `cif` predictions: https://github.com/rdk/p2rank/releases/tag/2.4.2

The build also includes a preview of new re-scoring model that doesn't use b-factor (`rescore_2024`). It's intended for both crystal structures and AlphaFold models (and all other types for structures: NMR, cryoEM,...). In all cases it should give better results than default re-scoring model, albeit it is ~3x slower. 

Usage:
`prank rescore -c rescore_2024 <dataset.ds>`

@Yorick-126 thank you for your notes. I have restored java 11 support for this version.
Please let me know if the new release and `rescore_2024` model work for you.

By the way, `alphafold` model is not supposed to be used in rescoring so I'm surprised that `./prank rescore -c ./config/alphafold` even works. 
(Note to myself: P2Rank should fail or at least warn users if they try to use wrong combiation of (command,config,model) )

**Comment by Yorick-126** [comment]
@rdk Thanks a lot for the updates!

Looking forward working with the new rescoring model, however I am not getting it to work. I downloaded and built your new version release (2.4.2) yesterday. Normal rescoring works just as fine as it did previously, e.g. `./distro/prank rescore /home/yorick/projectinstalls/p2rank/p2rank/distro/test_data/dataset_large.ds`

However, the new rescoring model `rescore_2024` doesn't work for me. It gives an error back that it failed to load the model rescore_2024 somehow. I am not sure what is happening there. Perhaps it is the java version? What version do you use?

The command I used is `./distro/prank rescore -c rescore_2024 /home/yorick/projectinstalls/p2rank/p2rank/distro/test_data/dataset_large.ds`

And this is the error I got:

```
[INFO] Dataset - loading dataset [/home/yorick/projectinstalls/p2rank/p2rank/distro/test_data/dataset_large.ds]
[INFO] Futils - deleting /home/yorick/projectinstalls/p2rank/p2rank/distro/test_output/rescore_dataset_large/run.log
rescoring pockets on proteins from dataset [dataset_large.ds]
[INFO] Console - rescoring pockets on proteins from dataset [dataset_large.ds]
[INFO] RescorePocketsRoutine - outdir: /home/yorick/projectinstalls/p2rank/p2rank/distro/test_output/rescore_dataset_large
[ERROR] Main - Failed to load model from file '/home/yorick/projectinstalls/p2rank/p2rank/distro/models/rescore_2024.model'
cz.siret.prank.program.PrankException: Failed to load model from file '/home/yorick/projectinstalls/p2rank/p2rank/distro/models/rescore_2024.model'
        at java.base/java.io.ObjectStreamClass.initNonProxy(ObjectStreamClass.java:597) ~[?:?]
        at java.base/java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:2051) ~[?:?]
        at java.base/java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1898) ~[?:?]
        at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2224) ~[?:?]
        at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1733) ~[?:?]
        at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:509) ~[?:?]
        at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:467) ~[?:?]
        at weka.core.SerializationHelper.read(SerializationHelper.java:278) ~[weka-dev-3.9.6.jar:?]
        at cz.siret.prank.utils.WekaUtils.loadClassifier(WekaUtils.groovy:86) ~[p2rank.jar:?]
        at cz.siret.prank.utils.WekaUtils.loadClassifier(WekaUtils.groovy:76) ~[p2rank.jar:?]
        at cz.siret.prank.program.ml.Model.loadFromFileV1(Model.groovy:90) ~[p2rank.jar:?]
        at cz.siret.prank.program.ml.Model.loadFromFile(Model.groovy:79) ~[p2rank.jar:?]
        at cz.siret.prank.program.ml.Model.load(Model.groovy:70) ~[p2rank.jar:?]
        at cz.siret.prank.program.routines.predict.RescorePocketsRoutine.execute(RescorePocketsRoutine.groovy:51) ~[p2rank.jar:?]
        at cz.siret.prank.program.Main.runRescore(Main.groovy:301) ~[p2rank.jar:?]
        at cz.siret.prank.program.Main.run(Main.groovy:400) ~[p2rank.jar:?]
        at cz.siret.prank.program.Main.main(Main.groovy:472) [p2rank.jar:?]
Caused by: java.io.InvalidClassException: cz.siret.prank.fforest.api.LegacyFlatBinaryForest; local class incompatible: stream classdesc serialVersionUID = -4570003757601764377, local class serialVersionUID = -1010621970731518626
        ... 17 more
ERROR: Failed to load model from file '/home/yorick/projectinstalls/p2rank/p2rank/distro/models/rescore_2024.model'
[ERROR] Console - Failed to load model from file '/home/yorick/projectinstalls/p2rank/p2rank/distro/models/rescore_2024.model'
For details see log file: /home/yorick/projectinstalls/p2rank/p2rank/distro/test_output/rescore_dataset_large/run.log
[INFO] Console - For details see log file: /home/yorick/projectinstalls/p2rank/p2rank/distro/test_output/rescore_dataset_large/run.log

[INFO] Console - 
Finished with ERROR in 0 hours 0 minutes 2.171 seconds.
[INFO] Console - Finished with ERROR in 0 hours 0 minutes 2.171 seconds.
```

java version is 17: 
```
openjdk version "17.0.11" 2024-04-16
OpenJDK Runtime Environment (build 17.0.11+9-Ubuntu-120.04.2)
OpenJDK 64-Bit Server VM (build 17.0.11+9-Ubuntu-120.04.2, mixed mode, sharing)
```

**Comment by rdk** [comment]
Are you using P2Rank version (a) downloaded from releases (https://github.com/rdk/p2rank/releases/download/2.4.2/p2rank_2.4.2.tar.gz) or (b) have you built it from sources from the current develop branch?
If (b) then please try (a).

**Comment by Yorick-126** [comment]
Ah thanks, that (a) worked! I indeed did (b).

**Comment by rdk** [comment]
It seems there is a bug on develop branch. Thanks for alerting me to it.

By the way, out of curiosity, is there any reason why you are rescoring predictions of fpocket rather than using pedictions of P2Rank (`prank predict`)? Or are you doing both?

**Comment by Yorick-126** [comment]
No problem!

Yeah sure! In our research department, we combine wet lab and computational methods. So for pocket identification we chose to focus on implementing fpocket in some of our workflows, because it is frequently used and also used as benchmark for other pocket predictors. I just recently learned of P2Rank, and based on your paper, it looks very promising. So for now, I am implementing the prank rescoring on fpocket in our workflows, and later on I'd like to incorporate P2Rank as well.
